foreign-types-shared/src/lib.rs

```diff
@@ -14,6 +14,9 @@ use core::mem;
 pub struct Opaque(PhantomData<UnsafeCell<*mut ()>>);
 
 /// A type implemented by wrappers over foreign types.
+///
+/// # Safety
+/// Any implementor of `ForeignType` must guarantee that `Self::from_ptr(x).to_ptr() == x`.
```

Comment:
There's a bit more that has to be guaranteed. Off the top of my head:

* `Self::from_ptr(x).into_ptr(x) == x`
* `Self::from_ptr(x).deref().as_ptr(x) == x`
* `Self::from_ptr(x).deref_mut().as_ptr(x) == x`
* `Self::from_ptr(x).as_ref().as_ptr(x) == x`
* `Self::from_ptr(x).as_mut().as_ptr(x) == x`

Reply:
Sure, added those.